0005-snarkyjs-ecdsa-sha.md

```diff
@@ -0,0 +1,197 @@
+
+For Keccak/SHA3, the implementation exposes two ready-to-go functions.
+
+```ocaml
+(** Gagdet for NIST SHA-3 function for output lengths 224/256/384/512 *)
+val nist_sha3 :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> int
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+(** Gadget for Keccak hash function for the parameters used in Ethereum *)
+val eth_keccak :
+    (module Snarky_backendless.Snark_intf.Run with type field = 'f)
+  -> 'f Snarky_backendless.Cvar.t list
+  -> 'f Snarky_backendless.Cvar.t array
+
+```
+
+These two functions will be imported into the bindings layer and exposed via a new sub-module as part of the already existing `Snarky` module.
+
+```ocaml
+
+  module Sha = struct
+    let create message nist length =
+      let message_array = Array.to_list message in
+      if Js.to_bool nist then
+        Kimchi_gadgets.Keccak.nist_sha3 (module Impl) length message_array
+      else Kimchi_gadgets.Keccak.eth_keccak (module Impl) message_array
+  end
+
+```
+
+In order to reduce the bindings surface, Keccak and SHA3 will be exposed via a `create` function, which behaves like a factory pattern.
+By calling this function inside SnarkyJS, we can define and expose all possible variants of SHA3(224/256/384/512) and Keccak without the need to have an individual function in the bindings layer for each variant.
+
+In SnarkyJS, these new primitives will be declared as followed:
+
+```ts
+function buildSha(length: 224 | 256 | 385 | 512, nist: boolean) {
+  return {
+    hash(message: Field[]) {
+      let payload = [0, ...message.map((f) => f.value)];
+      return Snarky.sha.create(payload, nist, length).map(Field);
+    },
+  };
+}
+
+const Sha3_224 = buildSha(224, true);
+const Sha3_256 = buildSha(256, true);
+const Sha3_385 = buildSha(385, true);
+const Sha3_512 = buildSha(512, true);
+const Keccak = buildSha(256, false);
+```
+
+Another alternative to the factory pattern above could be to supply the developer with a single function that takes a range of parameters, so that the developer can choose their flavour of SHA3/Keccak on their own.
+
+```ts
+function SHA3(
+  length: 224 | 256 | 385 | 512,
+  nist?: boolean = true
+): { hash(xs: Field[]) };
+```
+
+Developers can then import these functions into their project via
+
+```ts
+import { Sha3_224, Sha3_256, Sha3_385, Sha3_512, Keccak } from "snarkyjs";
+
+Sha3_224.hash(xs);
+// ..
+```
+
+### Alternative Approach
+
+Another possibility is to combine all hash functions, including Poseidon, under a shared namespace `Hash`. Developers will then be able to use these functions by calling `Hash.[hash_name].hash(xs)`. However, this would not be equivalent to the existing `Poseidon` API.
+
+However, the existing `Poseidon` API could be deprecated in favour of the new `Hash` namespace.
```

Comment:
Given that others prefer this approach, I'm ok with moving `Poseidon` to a new `Hash` namespace. I agree with Brandon that this probably makes it easier to find.

The concern I have is that such an interface does not single out Poseidon among all the hash functions. To the contrary, since `SHA3` and `Keccak` are more familiar, non-expert developers could default to one of these instead of Poseidon. However, Poseidon as the hash function that work on native field elements, and which our proof system is specifically optimized for, should clearly be the default choice. It takes 11 constraints where Keccak takes many thousands.

To single out Poseidon as _the_ recommended hash function, one approach could be to also add a top level `hash()` method:

```ts
import { hash, Hash } from "snarkyjs"

// the default hash function (= Poseidon)
hash([x,y,z]);

// other hash function for special needs and more advanced users
Hash.Keccak.hash([x,y,z]);
```

This way, we get a clear default hash function while removing the "Poseion" jargon (we can still explain what hash function this is in the doccomment)

Reply:
+1

Reply:
I like the idea of making `Poseidon` the default but having both a `hash` and `Hash` in SnarkyJS seems a little confusing although a developer can get more info with intellisense.

Reply:
We could have the `Poseidon` hash function set as the default hash on the `Hash` object/namespace, like this:

```ts
import { Hash } from "snarkyjs"
// It can be invoked like this
Hash.default([x,y,z]);
// Or can be this
Hash.hash([x,y,z]);
```

This way, there's no need to deal with multiple top-level imports for `hash` and `Hash`, which could potentially lead to confusion. Personally, I'm finding the idea of using `default` as the terminology is quite appealing, as it's clear and straightforward.

Reply:
or maybe just
```ts
Hash([x,y,z])
```
although I don't like upper case for a function that's not a constructor

Reply:
I like this option over the other two. 

```ts
// Or can be this
Hash.hash([x,y,z]);
```

Reply:
I'm not a fan of `Hash.default()`, but I'm ok with `Hash.hash()` even though it's a bit redundant. Personally I still prefer `hash()`

Reply:
How about `Hash.digest`? It breaks the `.hash` pattern but also signals uniqueness

Reply:
I agree it's nice to have a default hash -- I also prefer `Hash.hash()` over `Hash.default()`. I'm mixed on the naked `hash()`.

I think we won't reach consensus so we need someone to make a decision. I think @nicc should decide these sorts of things until we have a PM on the team.